### ISSUE TITLE
SSL test failures with OTP 23 and PostgreSQL 12.3

### ISSUE THREAD

**Issue by galdor** [epgsql/epgsql#238]
The test suite fails on Linux with OTP 23 and PostgreSQL 12.3:

```
%%% epgsql_SUITE ==> epgsql.connect.connect_with_invalid_client_cert: FAILED
%%% epgsql_SUITE ==>
Failure/Error: ?assertMatch({ error , { ssl_negotiation_failed , _ } }, Module : connect ( # { username => "epgsql_test_cert" , database => "epgsql_test_db1" , host => Host , port => Port , ssl => t
rue , ssl_opts => [ { keyfile , File ( "bad-client.key" ) } , { certfile , File ( "bad-client.crt" ) } ] } ))
  expected: = { error , { ssl_negotiation_failed , _ } }
       got: {error,
                {sock_error,
                    {tls_alert,
                        {unknown_ca,
                            "TLS client: In state connection received SERVER ALERT: Fatal - Unknown CA\n "}}}}
      line: 377
```

```
%%% epgsql_SUITE ==> epgsql_cast.connect.connect_with_invalid_client_cert: FAILED
%%% epgsql_SUITE ==>
Failure/Error: ?assertMatch({ error , { ssl_negotiation_failed , _ } }, Module : connect ( # { username => "epgsql_test_cert" , database => "epgsql_test_db1" , host => Host , port => Port , ssl => t
rue , ssl_opts => [ { keyfile , File ( "bad-client.key" ) } , { certfile , File ( "bad-client.crt" ) } ] } ))
  expected: = { error , { ssl_negotiation_failed , _ } }
       got: {error,
                {sock_error,
                    {tls_alert,
                        {unknown_ca,
                            "TLS client: In state connection received SERVER ALERT: Fatal - Unknown CA\n "}}}}
      line: 377

```

```
%%% epgsql_SUITE ==> epgsql_incremental.connect.connect_with_invalid_client_cert: FAILED
%%% epgsql_SUITE ==>
Failure/Error: ?assertMatch({ error , { ssl_negotiation_failed , _ } }, Module : connect ( # { username => "epgsql_test_cert" , database => "epgsql_test_db1" , host => Host , port => Port , ssl => true , ssl_opts => [ { keyfile , File ( "bad-client.key" ) } , { certfile , File ( "bad-client.crt" ) } ] } ))
  expected: = { error , { ssl_negotiation_failed , _ } }
       got: {error,
                {sock_error,
                    {tls_alert,
                        {unknown_ca,
                            "TLS client: In state connection received SERVER ALERT: Fatal - Unknown CA\n "}}}}
      line: 377
```

Since the test suite passes with OTP 23 and PostgreSQL 10 (see #237), there might be something different between Pg 10 and 12.

**Comment by seriyps** [comment]
Interesting. I tried to run epgsql test suite with my stock Ubuntu 20.04 postgres 12.2-4 - there were no issues. Do you think some incompatibilities were introduced in 12.3? Or maybe openssl? I have libssl 1.1.1f-1ubuntu2

**Comment by galdor** [comment]
Isn't Erlang using its own SSL implementation anyway?

I will post any information I can find if I find some time to investigate.

**Comment by seriyps** [comment]
`SERVER ALERT: Fatal - Unknown CA` means that it's server (PostgreSQL) is sending this alert to Erlang.
We indeed are using self-signed client-side certificates (see `generate_test_certs.sh`) to authenticate client. Maybe the format in which server reacts to wrong client certificate have changed in openssl or postgresql 12.3

**Comment by seriyps** [comment]
Hi. I managed to reproduce it with the new CI and it's most likely fixed in our tests (but still may happen in production if you use client-certificate TLS authentication.
Fixed by disabling TLS-1.3. Did not yet investigate why, but it seems Erlang's TLS-1.3 implementation sends the TLS alert asynchronously. So, initially TLS connection is reported as established and then instantly we get some asynchronous error.
https://github.com/epgsql/epgsql/pull/263/files#diff-00e608ddd8cbe7884b2ff01e68bc8f4bb750fb55183141391b2ec6dbad9893adR402
